### PR TITLE
Add `throwPluginError` to Plugin Utilities

### PR DESCRIPTION
**Pull request by drsooch**
Add the function `throwPluginError`. This function is intended to
provide a common `ResponseError` message for use in logging.

Renamed `response` to `pluginResponse` in `PluginUtils` for more clarity.

Shifting away from using `descriptor :: PluginID -> PluginDescriptor`. Instead the plugin ID is pulled out and defined as part of a plugin directly. This provides easy access to the PluginID rather than having to feed it through functions. 

I used the `callHierarchyPlugin` as a pseudo-testing ground for how it would look in practice. There will be a follow up PR that actually makes use of the error message (see #2919). 

I've also updated plugins I wrote to use the new descriptor model.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2924"></a>

### CODE REVIEW

File: hls-plugin-api/src/Ide/PluginUtils.hs [hunk 1]
```diff
@@ -255,13 +256,18 @@ getNormalizedFilePath (PluginId plId) docId = handleMaybe errMsg
...
 handleMaybe :: Monad m => e -> Maybe b -> ExceptT e m b
 handleMaybe msg = maybe (throwE msg) return
 
 handleMaybeM :: Monad m => e -> m (Maybe b) -> ExceptT e m b
 handleMaybeM msg act = maybeM (throwE msg) return $ lift act
 
-response :: Monad m => ExceptT String m a -> m (Either ResponseError a)
-response =
+pluginResponse :: Monad m => ExceptT String m a -> m (Either ResponseError a)
+pluginResponse =
```
Comment (drsooch) [inline hls-plugin-api/src/Ide/PluginUtils.hs:271]: Is there any gain to swap `ExceptT String m a` to `ExceptT Text m a`? I briefly experimented with it and it ripples up throughout most plugins, but not sure if it's even worth it?

**Review by kokobd — COMMENTED** [review]
LGTM. And I will update selection range plugin accordingly after this PR is merged.

**Review by michaelpj — APPROVED** [review]
Looks fine.

File: hls-plugin-api/src/Ide/PluginUtils.hs (hunk 1 above)
Comment (drsooch) [reply hls-plugin-api/src/Ide/PluginUtils.hs:270]: Figured as much. Maybe I'll look into it down the line.

File: hls-plugin-api/src/Ide/PluginUtils.hs [hunk 2]
```diff
@@ -254,13 +255,18 @@ getNormalizedFilePath (PluginId plId) uri = handleMaybe errMsg
         errMsg = T.unpack $ "Error(" <> plId <> "): converting " <> getUri uri <> " to NormalizedFilePath"
 
 -- ---------------------------------------------------------------------
+throwPluginError :: Monad m => PluginId -> String -> String -> ExceptT String m b
```
[the remark this answers is not in the document]
Comment (drsooch) [reply hls-plugin-api/src/Ide/PluginUtils.hs:258]: The intention was for users to give a short blurb of where it went wrong. I guess `HasCallStack` would work, but this is less about exception handling and more about: "The plugin has reached an unattainable state" or "The plugin can't progress" something along those lines. 

Let me play around with that idea and report back.

File: plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs [hunk 3]
```diff
@@ -144,14 +146,14 @@ p `isInsideRealSrcSpan` r = let (Range sp ep) = realSrcSpanToRange r in sp <= p
 
 getFirstPragma :: MonadIO m => IdeState -> NormalizedFilePath -> ExceptT String m NextPragmaInfo
 getFirstPragma state nfp = handleMaybeM "Error: Could not get NextPragmaInfo" $ do
-      ghcSession <- liftIO $ runAction "AlternateNumberFormat.GhcSession" state $ useWithStale GhcSession nfp
-      (_, fileContents) <- liftIO $ runAction "AlternateNumberFormat.GetFileContents" state $ getFileContents nfp
+      ghcSession <- liftIO $ runAction (alternateNumberFormatId <> ".GhcSession") state $ useWithStale GhcSession nfp
```
[the remark this answers is not in the document]
Comment (drsooch) [reply plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs:149]: Yeah #2919 I'm going to look to enhance the `pluginRecorder` to hopefully be able to decorate it with the plugin id. The other question is how to integrate that with `runAction` logging.

This one of the big things I want to look at considering all plugins have this same model of `id.Action`.

File: exe/Plugins.hs [hunk 4]
```diff
@@ -183,13 +183,13 @@ idePlugins recorder includeExamples = pluginDescToIdePlugins allPlugins
       Splice.descriptor "splice" :
 #endif
 #if alternateNumberFormat
-      AlternateNumberFormat.descriptor pluginRecorder "alternateNumberFormat" :
+      AlternateNumberFormat.descriptor pluginRecorder :
```
Comment (drsooch) [inline exe/Plugins.hs:186]: @wz1000 for reference this is what I was discussing in IRC. It seems like there are no dupe checks anyway so this should be a safe change?

File: plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs:149]: Yeah, I guess what I'm wondering is whether we want some more generic "log context" stack in our recorders or something. It comes up a lot.

**Comment by drsooch** [comment]
@michaelpj I've been thinking about this more. Would it make sense to just log the error within `throwPluginError`. This saves some follow up work with #2919. 

Something like this:

```haskell
throwPluginError' :: (HasCallStack, MonadIO m) => Recorder (WithPriority String) -> PluginId -> String -> String -> ExceptT String m b
throwPluginError' r (PluginId who) what _where' = logWith r Warning msg >> throwE msg
```

**Comment by drsooch** [comment]
> @michaelpj I've been thinking about this more. Would it make sense to just log the error within `throwPluginError`. This saves some follow up work with #2919.
> 
> Something like this:
> 
> ```haskell
> throwPluginError' :: (HasCallStack, MonadIO m) => Recorder (WithPriority String) -> PluginId -> String -> String -> ExceptT String m b
> throwPluginError' r (PluginId who) what _where' = logWith r Warning msg >> throwE msg
> ```

Now that I actually played around with it, I forgot we'd need to depend on `ghcide`, which I assume we don't want for `hls-plugins-api`

File: plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (July541) [reply plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs:151]: I strongly want something that can wrap `runAction` and return `ExceptT String m ActionResult`, then we can clean much boilerplate code.

**Comment by michaelpj** [comment]
> I've been thinking about this more. Would it make sense to just log the error within throwPluginError.

That's basically what I was thinking about. The `ghcide` dependency issue is annoying, though. I wonder if some of the logging stuff should live in `hls-plugin-api`? It's not inconceivable that we might want to e.g. specify that some plugin types get loggers, or allow plugins to log without depending on `ghcide`. It could just be the types and the functions for actually performing logging, the code for actually creating recorders could live in `ghcide` still? Or maybe we need the `hls-plugin-utils` package that @July541 has been thinking about.

**Pull request merged**